File: src/retrieval/service/merger.py
"""
Candidate Merger - RRF (Reciprocal Rank Fusion)
Deduplicate, score fusion, reranking, and safety filtering for three-way recall results.
Aligned with week-1 design doc.
"""

import logging
from typing import List, Dict
from datetime import datetime
from .types import Candidate, CandidateList, QueryMeta, RetrievalRequest

logger = logging.getLogger(__name__)


class Merger:
    """
    RRF Candidate Merger
    
    RRF_score(d) = Σ 1 / (k + rank_i(d))
    where k=60 (standard value), rank_i(d) is the rank of document d in path i
    """
    
    def __init__(
        self,
        k: int = 60,                    # RRF parameter
        min_confidence: float = 0.0,    # minimum confidence threshold
        seed_threshold: float = 0.7,    # score threshold for marking as seed
    ):
        self.k = k
        self.min_confidence = min_confidence
        self.seed_threshold = seed_threshold
    
    def merge(
        self,
        dense_results: List[Candidate],
        sparse_results: List[Candidate],
        filter_results: List[Candidate],
        request: RetrievalRequest,
    ) -> CandidateList:
        """
        Merge three-way retrieval results.
        
        Flow (aligned with week-1 design doc):
        1. Deduplicate by object_id, accumulate RRF scores
        2. Safety filter (quarantine, ttl, visible_time, is_active)
        3. Filter low confidence / low importance
        4. Rerank: final_score = rrf_score * importance * freshness_score * confidence
           If policy_records has confidence_override, it replaces confidence.
           salience_weight from policy_records applied as multiplier.
        5. Sort by final_score descending
        6. Truncate to effective top_k (top_k*2 if for_graph)
        7. Mark seed candidates (for member C)
        """
        start_time = datetime.now()
        
        # Step 1: Merge to map, accumulate RRF scores
        merged: Dict[str, Candidate] = {}
        
        self._add_with_rrf(merged, dense_results, "dense")
        self._add_with_rrf(merged, sparse_results, "sparse")
        self._add_with_rrf(merged, filter_results, "filter")
        
        candidates = list(merged.values())
        
        # Step 2: Safety filter
        candidates = self._safety_filter(candidates, request)
        
        # Step 3: Filter low confidence and low importance
        min_conf = request.min_confidence if request.min_confidence > 0 else self.min_confidence
        if min_conf > 0:
            candidates = [c for c in candidates if c.confidence >= min_conf]
        if request.min_importance > 0:
            candidates = [c for c in candidates if c.importance >= request.min_importance]
        
        # Step 4: Rerank with design doc formula
        for c in candidates:
            c.final_score = (c.score
                           * max(c.importance, 0.01)
                           * max(c.freshness_score, 0.01)
                           * max(c.confidence, 0.01))
        
        # Step 5: Sort by final_score descending
        candidates.sort(key=lambda c: c.final_score, reverse=True)
        
        # Step 6: Truncate - for_graph mode returns top_k*2
        top_k = request.top_k if request.top_k > 0 else 10
        effective_k = top_k * 2 if request.for_graph else top_k
        candidates = candidates[:effective_k]
        
        # Step 7: Mark seed candidates (for member C graph expansion)
        for c in candidates:
            if c.final_score >= self.seed_threshold:
                c.is_seed = True
                c.seed_score = c.final_score
        
        # Build metadata
        latency_ms = int((datetime.now() - start_time).total_seconds() * 1000)
        channels_used = []
        if dense_results:
            channels_used.append("dense")
        if sparse_results:
            channels_used.append("sparse")
        if filter_results:
            channels_used.append("filter")
        
        query_meta = QueryMeta(
            latency_ms=latency_ms,
            dense_hits=len(dense_results),
            sparse_hits=len(sparse_results),
            filter_hits=len(filter_results),
            channels_used=channels_used,
        )
        
        return CandidateList(
            candidates=candidates,
            total_found=len(merged),
            retrieved_at=datetime.now(),
            query_meta=query_meta,
        )
    
    def _add_with_rrf(
        self,
        merged: Dict[str, Candidate],
        results: List[Candidate],
        channel: str,
    ) -> None:
        """Calculate RRF score and accumulate to merged, track per-channel scores"""
        if not results:
            return
        
        for rank, candidate in enumerate(results, start=1):
            rrf_score = 1.0 / (self.k + rank)
            
            if candidate.object_id in merged:
                existing = merged[candidate.object_id]
                existing.score += rrf_score
                if channel not in existing.source_channels:
                    existing.source_channels.append(channel)
                # Track per-channel score on the merged candidate
                if channel == "dense":
                    existing.dense_score = rrf_score
                elif channel == "sparse":
                    existing.sparse_score = rrf_score
            else:
                candidate.score = rrf_score
                candidate.source_channels = [channel]
                if channel == "dense":
                    candidate.dense_score = rrf_score
                elif channel == "sparse":
                    candidate.sparse_score = rrf_score
                merged[candidate.object_id] = candidate
    
    def _safety_filter(self, candidates: List[Candidate], request: RetrievalRequest) -> List[Candidate]:
        """
        Post-merge safety filter (design doc section 4.5).
        Remove: quarantined, ttl expired, not yet visible, inactive.
        """
        now = datetime.now()
        filtered = []
        for c in candidates:
            if c.quarantine_flag:
                continue
            if c.ttl and c.ttl < now:
                continue
            if c.visible_time and c.visible_time > now:
                continue
            if not c.is_active:
                continue
            # as_of_ts: time-travel filter
            if request.as_of_ts:
                if c.visible_time and c.visible_time > request.as_of_ts:
                    continue
                if c.valid_from and c.valid_from > request.as_of_ts:
                    continue
            # min_version filter
            if request.min_version is not None and c.version < request.min_version:
                continue
            filtered.append(c)
        return filtered

File: cpp/retrieval/merger.cpp
// Copyright 2024 CogDB Authors
// SPDX-License-Identifier: Apache-2.0
//
// RRF merge and reranking implementation.

#include "andb/merger.h"
#include <algorithm>
#include <unordered_map>

namespace andb {

Merger::Merger(const MergeConfig& config) : config_(config) {}

Merger::~Merger() = default;

void Merger::SetConfig(const MergeConfig& config) {
    config_ = config;
}

const MergeConfig& Merger::GetConfig() const {
    return config_;
}

float Merger::ComputeRRFScore(int32_t rank) const {
    return 1.0f / static_cast<float>(config_.rrf_k + rank);
}

std::vector<Candidate> Merger::Merge(
    const SearchResult& dense_results,
    const SearchResult& sparse_results,
    int32_t top_k,
    bool for_graph
) const {
    // Step 1: Convert search results to candidates with RRF scores
    std::vector<Candidate> all_candidates;
    
    // Process dense results
    for (size_t i = 0; i < dense_results.ids.size(); ++i) {
        Candidate c;
        c.internal_id = dense_results.ids[i];
        c.dense_score = ComputeRRFScore(static_cast<int32_t>(i + 1));
        c.from_dense = true;
        all_candidates.push_back(c);
    }
    
    // Process sparse results
    for (size_t i = 0; i < sparse_results.ids.size(); ++i) {
        Candidate c;
        c.internal_id = sparse_results.ids[i];
        c.sparse_score = ComputeRRFScore(static_cast<int32_t>(i + 1));
        c.from_sparse = true;
        all_candidates.push_back(c);
    }
    
    // Step 2: Deduplicate by internal_id, accumulating RRF scores
    std::vector<Candidate> merged = Deduplicate(all_candidates);
    
    // Step 3: Compute RRF score (sum of per-channel scores)
    for (auto& c : merged) {
        c.rrf_score = c.dense_score + c.sparse_score;
    }
    
    // Step 4: Apply reranking formula
    Rerank(merged);
    
    // Step 5: Sort by final_score descending
    std::sort(merged.begin(), merged.end(),
        [](const Candidate& a, const Candidate& b) {
            return a.final_score > b.final_score;
        });
    
    // Step 6: Truncate to effective top_k
    int32_t effective_k = for_graph ? top_k * 2 : top_k;
    if (static_cast<int32_t>(merged.size()) > effective_k) {
        merged.resize(effective_k);
    }
    
    // Step 7: Mark seeds
    MarkSeeds(merged);
    
    return merged;
}

void Merger::Rerank(std::vector<Candidate>& candidates) const {
    for (auto& c : candidates) {
        c.final_score = ComputeFinalScore(
            c.rrf_score,
            c.importance,
            c.freshness_score,
            c.confidence
        );
    }
}

void Merger::MarkSeeds(std::vector<Candidate>& candidates) const {
    for (auto& c : candidates) {
        if (c.final_score >= config_.seed_threshold) {
            c.is_seed = true;
            c.seed_score = c.final_score;
        } else {
            c.is_seed = false;
            c.seed_score = 0.0f;
        }
    }
}

std::vector<Candidate> Merger::Deduplicate(
    const std::vector<Candidate>& candidates
) const {
    std::unordered_map<int64_t, Candidate> merged_map;
    
    for (const auto& c : candidates) {
        auto it = merged_map.find(c.internal_id);
        if (it == merged_map.end()) {
            merged_map[c.internal_id] = c;
        } else {
            // Accumulate scores
            it->second.dense_score += c.dense_score;
            it->second.sparse_score += c.sparse_score;
            it->second.from_dense = it->second.from_dense || c.from_dense;
            it->second.from_sparse = it->second.from_sparse || c.from_sparse;
            it->second.from_filter = it->second.from_filter || c.from_filter;
        }
    }
    
    std::vector<Candidate> result;
    result.reserve(merged_map.size());
    for (auto& [id, candidate] : merged_map) {
        result.push_back(std::move(candidate));
    }
    
    return result;
}

}  // namespace andb
